src/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.IO;
+using Internal.Runtime.Augments;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Threading
+{
+    public abstract partial class WaitHandle
+    {
+        private static bool WaitOneCore(IntPtr handle, int millisecondsTimeout)
+        {
+            Debug.Assert(millisecondsTimeout >= -1);
+
+            int ret = LowLevelThread.WaitForSingleObject(handle, millisecondsTimeout);
+
+            if (ret == WaitAbandoned)
+            {
+                ThrowAbandonedMutexException();
+            }
+
+            return ret != WaitTimeout;
+        }
+
+        /*========================================================================
+        ** Waits for signal from all the objects. 
+        ** timeout indicates how long to wait before the method returns.
+        ** This method will return either when all the object have been pulsed
+        ** or timeout milliseonds have elapsed.
+        ========================================================================*/
+        private static int WaitMultiple(
+            RuntimeThread currentThread,
+            SafeWaitHandle[] safeWaitHandles,
+            int count,
+            int millisecondsTimeout,
+            bool waitAll)
+        {
+            Debug.Assert(currentThread == RuntimeThread.CurrentThread);
+            Debug.Assert(safeWaitHandles != null);
+            Debug.Assert(safeWaitHandles.Length >= count);
+
+            IntPtr[] handles = currentThread.GetWaitedHandleArray(count);
+            for (int i = 0; i < count; i++)
+            {
+                handles[i] = safeWaitHandles[i].DangerousGetHandle();
```

Comment:
The addrefs are done in WaitHandle.cs just before calling the platform-specific code